src/odometry_2d/publisher.cpp

```diff
@@ -123,37 +123,45 @@ void Publisher::notifyCallback(
-    odom_output_.pose.covariance[35] = covariance_matrices[2][0];
-
-    odom_output_.twist.covariance[0] = covariance_matrices[3][0];
-    odom_output_.twist.covariance[1] = covariance_matrices[3][1];
-    odom_output_.twist.covariance[5] = covariance_matrices[4][0];
-    odom_output_.twist.covariance[6] = covariance_matrices[3][2];
-    odom_output_.twist.covariance[7] = covariance_matrices[3][3];
-    odom_output_.twist.covariance[11] = covariance_matrices[4][1];
-    odom_output_.twist.covariance[30] = covariance_matrices[4][0];
-    odom_output_.twist.covariance[31] = covariance_matrices[4][1];
-    odom_output_.twist.covariance[35] = covariance_matrices[5][0];
-
+    try
+    {
+      std::vector<std::pair<fuse_core::UUID, fuse_core::UUID>> covariance_requests;
+      covariance_requests.emplace_back(position_uuid, position_uuid);
+      covariance_requests.emplace_back(position_uuid, orientation_uuid);
+      covariance_requests.emplace_back(orientation_uuid, orientation_uuid);
+      covariance_requests.emplace_back(velocity_linear_uuid, velocity_linear_uuid);
+      covariance_requests.emplace_back(velocity_linear_uuid, velocity_angular_uuid);
+      covariance_requests.emplace_back(velocity_angular_uuid, velocity_angular_uuid);
+
+      std::vector<std::vector<double>> covariance_matrices;
+      graph->getCovariance(covariance_requests, covariance_matrices);
+
+      odom_output_.pose.covariance[0] = covariance_matrices[0][0];
+      odom_output_.pose.covariance[1] = covariance_matrices[0][1];
+      odom_output_.pose.covariance[5] = covariance_matrices[1][0];
+      odom_output_.pose.covariance[6] = covariance_matrices[0][2];
+      odom_output_.pose.covariance[7] = covariance_matrices[0][3];
+      odom_output_.pose.covariance[11] = covariance_matrices[1][1];
+      odom_output_.pose.covariance[30] = covariance_matrices[1][0];
+      odom_output_.pose.covariance[31] = covariance_matrices[1][1];
+      odom_output_.pose.covariance[35] = covariance_matrices[2][0];
+
+      odom_output_.twist.covariance[0] = covariance_matrices[3][0];
+      odom_output_.twist.covariance[1] = covariance_matrices[3][1];
+      odom_output_.twist.covariance[5] = covariance_matrices[4][0];
+      odom_output_.twist.covariance[6] = covariance_matrices[3][2];
+      odom_output_.twist.covariance[7] = covariance_matrices[3][3];
+      odom_output_.twist.covariance[11] = covariance_matrices[4][1];
+      odom_output_.twist.covariance[30] = covariance_matrices[4][0];
+      odom_output_.twist.covariance[31] = covariance_matrices[4][1];
+      odom_output_.twist.covariance[35] = covariance_matrices[5][0];
+    }
+    catch (const std::exception& e)
+    {
+      ROS_FATAL_STREAM("An error occurred computing the covariance information for " << latest_stamp_ << ":\n" <<
+                       e.what());
+    }
+    // TODO(swilliams): Should we always publish the odometry message, even if the covariance information is missing?
```

Comment:
This will publish the last good covariance. If we're going to pubish when the covariance computation fails, we should set the `odom_output_` covariances to all zeros (or any invalid covariance that can be detected in the consumer side).

I lean towards publishing even if the covariance couldn't be computed. In that case, I'd do:
1. When the covariance computation fails, set the `odom_output_.pose.covariance` or `odom_output_.twist.covariance` to all zeros. Any consumer should check the covariance is valid, so that way it's possible to detect the covariance couldn't be computed and either use the pose or disregard it.
2. Request the pose and twist covariance in two separate `try`-`catch` blocks. I guess there are cases where only one of them can't be computed.

In theory, as you're saying, if the covariance couldn't be computed, it means that something is wrong with the problem we're trying to optimize and the solution found is likely wrong. I think it's better to still publish and explicitly tell the consumer the problem couldn't be solve (or at least the covariance), as opposed to not publishing, which could also mean that something else happened (e.g. it took too long to publish, a configuration issue, ...).

That being said, if I had to implement a consumer of this output, TBH I wouldn't use the `pose` or `twist` if their covariance couldn't be computed. In the best case, it might just be that the covariance is rank deficient, so likely some components of the `pose` or `twist` are close to the optimal, but there are others that aren't, but even in that case I don't see myself using only some `pose` or `twist` components.

Reply:
1. Clearing the covariance is definitely a good call.
2. I'm somewhat hesitant to separate the covariance computation into two operations. This is a heavy computation, and is much more efficient to perform as a single operation rather than two. I'm also unsure if the current "rank deficient" error is affected by which covariance blocks are requested. I'll do some digging in the Ceres code. If the "rank deficient" error will occur regardless of what covariance blocks are requested, I am going to leave the computation as a single operation. If the covariance block request might affect the "rank deficient" error, we can discuss whether breaking them into two operations is worth the performance hit or not.

Reply:
2. Ceres forms a (usually sparse) representation of the entire Jacobian, then performs a QR decomposition on the full Jacobian to determine the rank. This will occur regardless what specific covariance blocks were requested. As such I'm going to leave the position and velocity covariance requests in a single call to getCovariance().